vortex-array/src/view.rs

```diff
@@ -136,8 +141,198 @@ impl<'v> ArrayView<'v> {
     }
 
     pub fn statistics(&self) -> &dyn Statistics {
-        // TODO(ngates): store statistics in FlatBuffers
-        &EmptyStatistics
+        self
+    }
+}
+
+impl Statistics for ArrayView<'_> {
+    fn get(&self, stat: Stat) -> Option<Scalar> {
+        // fb fetch is just a pointer dereference, so we check that first
+        let from_fb = get_from_flatbuffer_array(self.array, stat);
+        if from_fb.is_some() {
+            return from_fb;
+        }
+        None
+    }
+
+    /// NB: part of the contract for to_set is that it does not do any expensive computation.
+    /// In other implementations, this means returning the underlying stats map, but for the flatbuffer
+    /// implemetation, we have 'precalculated' stats in the flatbuffer itself, so we need to
+    /// alllocate a stats map and populate it with those fields.
+    fn to_set(&self) -> StatsSet {
+        let mut result = StatsSet::new();
+        for stat in all::<Stat>() {
+            if let Some(value) = self.get(stat) {
+                result.set(stat, value)
+            }
+        }
+        result
+    }
+
+    /// We want to avoid any sort of allocation on instantiation of the ArrayView, so we
+    /// do not allocate a stats_set to cache values.
+    fn set(&self, _stat: Stat, _value: Scalar) {
+        unimplemented!()
+    }
+
+    fn compute(&self, stat: Stat) -> Option<Scalar> {
+        if let Some(s) = self.get(stat) {
+            return Some(s);
+        }
+
+        let calculated = self
+            .to_array()
+            .with_dyn(|a| a.compute_statistics(stat))
+            .ok()?;
+
+        calculated.into_iter().for_each(|(k, v)| self.set(k, v));
+        self.get(stat)
+    }
+
+    fn with_stat_value<'a>(
+        &self,
+        stat: Stat,
+        f: &'a mut dyn FnMut(&Scalar) -> VortexResult<()>,
+    ) -> VortexResult<()> {
+        if let Some(existing) = self.get(stat) {
+            return f(&existing);
+        }
+        vortex_bail!(ComputeError: "statistic {} missing", stat);
+    }
+
+    fn with_computed_stat_value<'a>(
+        &self,
+        stat: Stat,
+        f: &'a mut dyn FnMut(&Scalar) -> VortexResult<()>,
+    ) -> VortexResult<()> {
+        self.compute(stat)
+            .map(|s| f(&s))
+            .unwrap_or_else(|| vortex_bail!(ComputeError: "statistic {} missing", stat))
+    }
+}
+
+fn get_from_flatbuffer_array(array: fb::Array<'_>, stat: Stat) -> Option<Scalar> {
```

Comment:
I'm not a fan of pulling out functions when there's only a single call site. Can just inline this into the get.

Reply:
Sure, will keep that in mind.